File: Scripts/utils/config_utils.js
/// <reference path="../js_typings/gctv.d.ts" />
/// <reference path="../js_typings/gtav_natives.d.ts" /> 

// Constants for configuration file names.
const FEATURE_SETTINGS_FILE = "feature_settings.json";
const GCT_CONFIG_FILE = "config.json";

/**
 * A utility object for managing configuration settings stored in JSON files.
 * Provides methods to get and set feature-specific and general GCTV settings.
 */
const configUtils = {
    /**
     * Internal helper to read JSON data from a file using the global `fs` module.
     * @param {string} filePath - The path to the JSON configuration file.
     * @returns {object} The parsed JSON data, or an empty object if the file
     * does not exist, is empty, or contains invalid JSON.
     */
    _readJsonFile(filePath) {
        if (!fs.fileExists(filePath)) {
            return {};
        }
        try {
            const data = fs.readFile(filePath);
            if (data === null || data === undefined || data === "") {
                return {}; // Return empty object if file is empty
            }
            return JSON.parse(data);
        } catch (e) {
            console.error(`Error: Could not read or decode JSON from ${filePath}. Returning empty config.`, e);
            return {};
        }
    },

    /**
     * Internal helper to save data to a JSON file using the global `fs` module.
     * @param {string} filePath - The path to the JSON configuration file.
     * @param {object} data - The JavaScript object to save as JSON.
     */
    _saveJsonFile(filePath, data) {
        try {
            // Convert the JavaScript object to a JSON string, pretty-printed with 4 spaces.
            const jsonString = JSON.stringify(data, null, 4);
            fs.writeFile(filePath, jsonString);
        } catch (e) {
            console.error(`Error: Could not save JSON to ${filePath}:`, e);
        }
    },

    /**
     * Retrieves a specific feature setting from the feature_settings.json file.
     * @param {string} section - The main section name (e.g., "Hotkeys").
     * @param {string} setting - The specific setting name within the section (e.g., "VehicleBoostKey").
     * @returns {any} The value of the setting, or undefined if not found.
     */
    getFeatureSetting(section, setting) {
        const featuresSettings = this._readJsonFile(FEATURE_SETTINGS_FILE);
        // Use optional chaining (?.) for safe access to nested properties.
        return featuresSettings[section]?.[setting];
    },

    /**
     * Sets a specific feature setting and saves it to the feature_settings.json file.
     * Creates the section or file if they do not exist.
     * @param {string} section - The main section name.
     * @param {string} setting - The specific setting name.
     * @param {any} value - The value to set.
     */
    setFeatureSetting(section, setting, value) {
        const featuresSettings = this._readJsonFile(FEATURE_SETTINGS_FILE);
        if (!featuresSettings[section]) {
            featuresSettings[section] = {}; // Create section if it doesn't exist
        }
        featuresSettings[section][setting] = value;
        this._saveJsonFile(FEATURE_SETTINGS_FILE, featuresSettings);
        console.log(`Feature setting '${section}.${setting}' set to '${value}'.`);
    },

    /**
     * Retrieves a general GCTV setting from the config.json file.
     * @param {string} setting - The specific setting name.
     * @returns {any} The value of the setting, or undefined if not found.
     */
    getGctSetting(setting) {
        const gctSettings = this._readJsonFile(GCT_CONFIG_FILE);
        return gctSettings[setting];
    },

    /**
     * Sets a general GCTV setting and saves it to the config.json file.
     * Creates the file if it does not exist.
     * @param {string} setting - The specific setting name.
     * @param {any} value - The value to set.
     */
    setGctSetting(setting, value) {
        const gctSettings = this._readJsonFile(GCT_CONFIG_FILE);
        gctSettings[setting] = value;
        this._saveJsonFile(GCT_CONFIG_FILE, gctSettings);
        console.log(`GCT setting '${setting}' set to '${value}'.`);
    }
};

/*
// Example Usage (commented out for production, uncomment for testing in your Duktape environment)
// This example assumes `fs` is globally available.
if (typeof fs !== 'undefined' && typeof console !== 'undefined') {
    console.log("--- Testing configUtils JavaScript (Duktape with fs) ---");

    // To ensure a clean test, you might need to manually delete the .json files
    // before running this example, as there's no `fs.clear()` equivalent.
    // Example: fs.deleteFile(FEATURE_SETTINGS_FILE); fs.deleteFile(GCT_CONFIG_FILE);

    // Set and get feature settings
    configUtils.setFeatureSetting("Hotkeys", "VehicleBoostKey", "F5");
    configUtils.setFeatureSetting("VehicleOptions", "AutoFixVehicle", true);
    const boostKey = configUtils.getFeatureSetting("Hotkeys", "VehicleBoostKey");
    const autoFix = configUtils.getFeatureSetting("VehicleOptions", "AutoFixVehicle");
    console.log(`Retrieved VehicleBoostKey: ${boostKey}`);
    console.log(`Retrieved AutoFixVehicle: ${autoFix}`);
    console.log(`Non-existent setting: ${configUtils.getFeatureSetting('NonExistent', 'Key')}`);

    // Set and get GCT settings
    configUtils.setGctSetting("GameVolume", 0.75);
    configUtils.setGctSetting("PlayerName", "TestPlayerJS");
    const volume = configUtils.getGctSetting("GameVolume");
    const playerName = configUtils.getGctSetting("PlayerName");
    console.log(`Retrieved GameVolume: ${volume}`);
    console.log(`Retrieved PlayerName: ${playerName}`);
    console.log(`Non-existent GCT setting: ${configUtils.getGctSetting('DebugMode')}`);

    console.log("Configuration files should now be updated on disk.");
} else {
    console.warn("`fs` module or `console` not available. Example usage will not run.");
}
*/

File: Scripts/utils/config_utils.py
import json
import os

# Constants for configuration file names.
FEATURE_SETTINGS_FILE = "feature_settings.json"
GCT_CONFIG_FILE = "config.json"

class ConfigUtils:
    """
    A utility class for managing configuration settings stored in JSON files.
    Provides methods to get and set feature-specific and general GCTV settings.
    """

    @staticmethod
    def _read_json_file(file_path):
        """
        Internal helper to read a JSON file.
        Returns an empty dictionary if the file does not exist or is empty/invalid.
        """
        if not os.path.exists(file_path) or os.path.getsize(file_path) == 0:
            return {}
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except json.JSONDecodeError:
            print(f"Error: Could not decode JSON from {file_path}. Returning empty config.")
            return {}
        except Exception as e:
            print(f"An unexpected error occurred while reading {file_path}: {e}")
            return {}

    @staticmethod
    def _save_json_file(file_path, data):
        """
        Internal helper to save data to a JSON file.
        """
        try:
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=4) # Use indent for pretty-printing
        except Exception as e:
            print(f"Error: Could not save JSON to {file_path}: {e}")

    @staticmethod
    def get_feature_setting(section, setting):
        """
        Retrieves a specific feature setting from the feature_settings.json file.
        
        Args:
            section (str): The main section name (e.g., "Hotkeys").
            setting (str): The specific setting name within the section (e.g., "VehicleBoostKey").
            
        Returns:
            any: The value of the setting, or None if not found.
        """
        features_settings = ConfigUtils._read_json_file(FEATURE_SETTINGS_FILE)
        return features_settings.get(section, {}).get(setting)

    @staticmethod
    def set_feature_setting(section, setting, value):
        """
        Sets a specific feature setting and saves it to the feature_settings.json file.
        Creates the section or file if they do not exist.
        
        Args:
            section (str): The main section name.
            setting (str): The specific setting name.
            value (any): The value to set.
        """
        features_settings = ConfigUtils._read_json_file(FEATURE_SETTINGS_FILE)
        if section not in features_settings:
            features_settings[section] = {}
        features_settings[section][setting] = value
        ConfigUtils._save_json_file(FEATURE_SETTINGS_FILE, features_settings)
        print(f"Feature setting '{section}.{setting}' set to '{value}'.")

    @staticmethod
    def get_gct_setting(setting):
        """
        Retrieves a general GCTV setting from the config.json file.
        
        Args:
            setting (str): The specific setting name.
            
        Returns:
            any: The value of the setting, or None if not found.
        """
        gct_settings = ConfigUtils._read_json_file(GCT_CONFIG_FILE)
        return gct_settings.get(setting)

    @staticmethod
    def set_gct_setting(setting, value):
        """
        Sets a general GCTV setting and saves it to the config.json file.
        Creates the file if it does not exist.
        
        Args:
            setting (str): The specific setting name.
            value (any): The value to set.
        """
        gct_settings = ConfigUtils._read_json_file(GCT_CONFIG_FILE)
        gct_settings[setting] = value
        ConfigUtils._save_json_file(GCT_CONFIG_FILE, gct_settings)
        print(f"GCT setting '{setting}' set to '{value}'.")

# Example Usage (for testing purposes, remove in production if not needed)
if __name__ == "__main__":
    print("--- Testing ConfigUtils Python ---")

    # Set and get feature settings
    ConfigUtils.set_feature_setting("Hotkeys", "VehicleBoostKey", "F5")
    ConfigUtils.set_feature_setting("VehicleOptions", "AutoFixVehicle", True)
    boost_key = ConfigUtils.get_feature_setting("Hotkeys", "VehicleBoostKey")
    auto_fix = ConfigUtils.get_feature_setting("VehicleOptions", "AutoFixVehicle")
    print(f"Retrieved VehicleBoostKey: {boost_key}")
    print(f"Retrieved AutoFixVehicle: {auto_fix}")
    print(f"Non-existent setting: {ConfigUtils.get_feature_setting('NonExistent', 'Key')}")

    # Set and get GCT settings
    ConfigUtils.set_gct_setting("GameVolume", 0.75)
    ConfigUtils.set_gct_setting("PlayerName", "TestPlayer")
    volume = ConfigUtils.get_gct_setting("GameVolume")
    player_name = ConfigUtils.get_gct_setting("PlayerName")
    print(f"Retrieved GameVolume: {volume}")
    print(f"Retrieved PlayerName: {player_name}")
    print(f"Non-existent GCT setting: {ConfigUtils.get_gct_setting('DebugMode')}")

    # Clean up (optional, for testing)
    # import os
    # if os.path.exists(FEATURE_SETTINGS_FILE):
    #     os.remove(FEATURE_SETTINGS_FILE)
    # if os.path.exists(GCT_CONFIG_FILE):
    #     os.remove(GCT_CONFIG_FILE)
    # print("Cleaned up test config files.")
